src/cookbooks/clients_last_seen_bits.md

```diff
@@ -0,0 +1,689 @@
+here's how this client looks:
+
+```
+SELECT
+  submission_date,
+  udf.bits28_retention(days_seen_bits, submission_date).day_13.*
+FROM
+  telemetry.clients_last_seen
+
+/*
+                   submission_date = 2020-01-28
+                       metric_date = 2020-01-15
+                   active_on_day_0 = true
+                  active_in_week_0 = true
+active_in_week_0_after_metric_date = false
+                  active_in_week_1 = true
+*/
+```
+
+### N-day Retention
+
+As an example of a novel metric that can be defined using the low-level
+bit pattern UDFs, let's define `n`-day retention as the fraction of clients active on a given day
+who are also active within the next `n` days. For example, 3-day retention would
+have a denominator of all clients active on day 0 and a numerator of all clients
+active on day 0 who were also active on days 1 or 2.
+
+To calculate `n`-day retention, we need to use the lower-level `bits28_range`
+function:
+
+```
+DECLARE n INT64;
+SET n = 3;
+
+WITH base AS (
+  SELECT
+    *,
+    udf.bits28_active_in_range(days_seen_bits, -n + 1, 1) AS seen_on_day_0,
+    udf.bits28_active_in_range(days_seen_bits, -n + 2, n - 1) AS seen_after_day_0
+  FROM
+    telemetry.clients_last_seen )
+SELECT
+  DATE_SUB(submission_date, INTERVAL n DAY) AS metric_date,
+  SAFE_DIVIDE(
+    COUNTIF(seen_on_day_0 AND seen_after_day_0),
+    COUNTIF(seen_on_day_0)
+  ) AS retention_n_day
+FROM
+  base
+WHERE
+  submission_date = '2020-01-28'
+GROUP BY
+  metric_date
+```
+
+### Retention using activity date
+
+GUD's canonical retention definitions are all based on ping submission dates rather
+than logical activity dates taken from client-provided timestamps, but there is
+interest in using client timestamps particular for `n`-day retention calculations
+for mobile products.
+
+Let's consider Firefox Preview which sends telemetry via Glean. The
+`org_mozilla_fenix.baseline_clients_last_seen` table includes two bit patterns
+that encode client timestamps: `days_seen_session_start_bits` and
+`days_seen_session_end_bits`. This table is still populated once per day based
+on pings received over the previous day, but some of those pings will reflect
+sessions that started on previous days. This introduces some new complexity
+into retention calculations because we'll always be underestimating client
+counts if we have our retention window end on `submission_date`.
+
+When using activity date, it may be desirable to build in a few days of buffer
+to ensure we are considering late-arriving pings. For example, if we wanted
+to calculate 3-day retention but allow 2 days of cushion for late-arriving
+pings, we would need to use an offset of 5 days from `submission_date`:
+
+```
+DECLARE n, cushion_days, offset_to_day_0 INT64;
+SET n = 3;
+SET cushion_days = 2;
+SET offset_to_day_0 = -n - cushion_days;
+
+WITH base AS (
+  SELECT
+    *,
+    udf.bits28_active_in_range(days_seen_session_start_bits, offset_to_day_0 + 0, 1) AS seen_on_day_0,
```

Comment:
Do we need to update this query as well?

    udf.bits28_active_in_range(days_seen_session_start_bits, offset_to_day_0+ 1, 1) AS seen_on_day_0,
    udf.bits28_active_in_range(days_seen_session_start_bits, offset_to_day_0 + 2, n - 1) AS seen_after_day_0